test/Fencer/Rules/Test.hs

```diff
@@ -181,6 +149,123 @@ test_rulesLoadRulesMinimal =
+-- presence of another configuration file without read permissions.
+--
+-- This matches the behavior of @lyft/ratelimit@.
+test_rulesLoadRulesReadPermissions :: TestTree
+test_rulesLoadRulesReadPermissions =
+  testCase "Configuration file read permissions" $
+    expectLoadRulesWithPermissions
+      (#ignoreDotFiles False)
+      (#files
+        [ ("domain1" </> "config.yml", domain1Text, const Dir.emptyPermissions)
+        , ("domain2" </> "config" </> "config.yml", domain2Text, id) ]
+      )
+      (#result $ Right [domain2])
+
+----------------------------------------------------------------------------
+-- Helpers
+----------------------------------------------------------------------------
+
+-- | Get a list of values on the Left or an empty list if it is a
+-- Right value.
+toErrorList :: Either [a] [b] -> [a]
+toErrorList (Right _) = []
+toErrorList (Left xs) = xs
+
+-- | Write contents to a path in the given root and modify file
+-- permissions.
+writeFile
+  :: "root" :! FilePath
+  -> "path" :! FilePath
+  -> "content" :! Text
+  -> "modifyPerms" :! (Dir.Permissions -> Dir.Permissions)
+  -> IO ()
+writeFile
+  (arg #root -> root)
+  (arg #path -> path)
+  (arg #content -> content)
+  (arg #modifyPerms -> modifyPerms) = do
+
+  let
+    dir = takeDirectory path
+    fullPath = root </> path
+  Dir.createDirectoryIfMissing True (root </> dir)
+  TIO.writeFile fullPath content
+  perms <- Dir.getPermissions fullPath
+  Dir.setPermissions fullPath (modifyPerms perms)
+
+-- | Write the content of files at the given root and load the files.
+writeAndLoadRules
+  :: "ignoreDotFiles" :! Bool
+  -> "root" :! FilePath
+  -> "files" :! [(FilePath, Text, Dir.Permissions -> Dir.Permissions)]
+  -> IO (Either [LoadRulesError] [DomainDefinition])
+writeAndLoadRules
+  (arg #ignoreDotFiles -> ignoreDotFiles)
+  (arg #root -> root)
+  (arg #files -> files) = do
+
+  forM_ files $ \(path, txt, permUpdate) -> Fencer.Rules.Test.writeFile
+    (#root root)
+    (#path path)
+    (#content txt)
+    (#modifyPerms permUpdate)
+  loadRulesFromDirectory
+    (#rootDirectory root)
+    (#subDirectory ".")
+    (#ignoreDotFiles ignoreDotFiles)
+
+-- | Create given directory structure and check that
+-- 'loadRulesFromDirectory' produces expected result such that file
+-- permissions are configurable.
+expectLoadRulesWithPermissions
+  :: "ignoreDotFiles" :! Bool
+  -> "files" :! [(FilePath, Text, Dir.Permissions -> Dir.Permissions)]
+  -> "result" :! Either [LoadRulesError] [DomainDefinition]
+  -> Assertion
+expectLoadRulesWithPermissions
+  (arg #ignoreDotFiles -> ignoreDotFiles)
+  (arg #files -> files)
+  (arg #result -> result) =
+  Temp.withSystemTempDirectory "fencer-config" $ \tempDir ->
+    writeAndLoadRules
+      (#ignoreDotFiles ignoreDotFiles)
+      (#root tempDir)
+      (#files files)
+      >>= \case
+      f@(Left _) ->
+        -- Paths to temporary files vary and there is not much point
```

Comment:
Do you mean that you can't compare exceptions because they include filepaths? Because otherwise I can't agree there isn't any point in specifying which exception messages are expected.

Reply:
I addressed this by comparing `LoadRuleError`s on their string representation modulo `FilePath`s that are trimmed to their file names (i.e., the directory paths are dropped).